File: 백준/Silver/15655. N과 M （6）/N과 M （6）.py
"""
조합 기본코드
"""

import sys

input = lambda: sys.stdin.readline().strip()

N, M = map(int, input().split())
lst = list(map(int, input().split()))

lst = sorted(lst)

selected = [0] * M


def recur(cur, start):
	if cur == M:
		for i in selected:
			print(lst[i], end=" ")
		print()
		return
	for i in range(start, N):
		selected[cur] = i
		recur(cur + 1, i + 1)


recur(0, 0)

File: 백준/Silver/15655. N과 M （6）/N과 M （6）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

/**
 * 조합 기본 코드
 */
public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static StringBuilder sb = new StringBuilder();
	private static int N;
	private static int M;
	private static int[] selected;
	private static int[] lst;
	
	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		selected = new int[M];
		lst = new int[N];
		
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < N; i++) {
			lst[i] = Integer.parseInt(st.nextToken());
		}
		
		Arrays.sort(lst);
		
		recur(0, 0);
		
		System.out.println(sb.toString());
	} // end of main

	private static void recur(int cur, int start) {
		if (cur == M) {
			for (int i : selected) {
				sb.append(lst[i] + " ");
			}
			sb.append("\n");
			return;
		}
		for (int i = start; i < N; i++) {
			selected[cur] = i;
			recur(cur + 1, i + 1);
		}
	} // end of recur
} // end of class